src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala

```diff
@@ -0,0 +1,161 @@
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      logger.debug(s"received sync message: $message")
+
+      val groupId: WorkbenchGroupIdentity = parseMessage(message)
+
+      googleExtensions.synchronizeGroupMembers(groupId).toTry.map(sr => (sr, message.ackId)) pipeTo self
+
+    case None =>
+      // there was no message to wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime.asInstanceOf[FiniteDuration], self, StartMonitorPass)
+
+    case (Success(report: SyncReport), ackId: String) =>
+      val errorReports = report.items.collect {
+        case SyncReportItem(_, _, errorReports) if errorReports.nonEmpty => errorReports
+      }.flatten
+
+      if (errorReports.isEmpty) {
+        // sync done, log it and try again immediately
+        acknowledgeMessage(ackId).map(_ => StartMonitorPass) pipeTo self
+
+        import DefaultJsonProtocol._
+        import org.broadinstitute.dsde.workbench.sam.google.GoogleModelJsonSupport._
+        logger.info(s"synchronized google group ${report.groupEmail.value}: ${report.items.toJson.compactPrint}")
+      } else {
+        throw new WorkbenchExceptionWithErrorReport(ErrorReport("error(s) syncing google group", errorReports))
+      }
+
+    case (Failure(t), ackId: String) =>
+      t match {
+        case groupNotFound: WorkbenchExceptionWithErrorReport if groupNotFound.errorReport.statusCode == Some(StatusCodes.NotFound) =>
+          // this can happen if a group is created then removed before the sync message is handled
+          // acknowledge it so we don't have to handle it again
+          acknowledgeMessage(ackId).map(_ => StartMonitorPass) pipeTo self
+          logger.info(s"group to synchronize not found: ${groupNotFound.errorReport}")
+
+        case regrets: Throwable => throw regrets
+      }
+
+    case Status.Failure(t) => throw t
+
+    case ReceiveTimeout =>
+      throw new WorkbenchException("GoogleGroupSyncMonitorActor has received no messages for too long")
+
+    case x => logger.info(s"unhandled $x")
+  }
+
+  private def acknowledgeMessage(ackId: String) = {
+    pubSubDao.acknowledgeMessagesById(pubSubSubscriptionName, Seq(ackId))
+  }
+
+  private def parseMessage(message: PubSubMessage) = {
+    (Try {
+      import org.broadinstitute.dsde.workbench.sam.model.SamJsonSupport.ResourceAndPolicyNameFormat
+      message.contents.parseJson.convertTo[ResourceAndPolicyName]
+    } recover {
+      case _: DeserializationException =>
+        import WorkbenchIdentityJsonSupport.WorkbenchGroupNameFormat
+        message.contents.parseJson.convertTo[WorkbenchGroupName]
```

Comment:
Is this to handle old-style name-only messages? Based on previous discussions, I thought sam was going to only pick up new-style dn messages and we'd let rawls handle any lingering old-style messages.

Reply:
I was originally thinking the messages would look like a ldap dn but changed my mind when it came time to implement because none of our DAO functions are in terms of dns. The message coming in could be either a serialized ResourceAndPolicyName or WorkbenchGroupName. The only way to figure it out is to try one then the other or make a more complicated json structure that indicates which kind it is.

Reply:
Or send to different topics which is what I was envisioning. I think this had some advantages when it came to rolling out sam and letting it handle any new messages while rawls could continue to drain the queue of old messages. Does publishing to the same queue complicate rollout? At the very least, rawls probably needs a change so it doesn't choke on new-style messages.

Reply:
we kept the different queue idea (it is all config based). Rawls submits to and consumes one and sam another.

Reply:
Okay. I was hoping to find a reason to avoid this try-one-then-the-other code but I guess it's just how it has to be.